### PR TITLE
Tests for New-WinEvent

### PR DESCRIPTION
**Pull request by anmenaga**
Some tests for New-WinEvent that bring code coverage of the class to 80%.

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,33 @@
+Describe 'New-WinEvent' -Tags "CI" {
+
+    Context "New-WinEvent tests" {
+        
+        It 'Simple New-WinEvent' -Skip:(-not $IsWindows) {
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: You may want to use $PSDefaultParameterValues to set the value of skip.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            New-WinEvent -ProviderName Microsoft-Windows-PowerShell -Id 40962 -Version 1 # simple event without any payload
+            $filter = @{ ProviderName = 'Microsoft-Windows-PowerShell'; Id = 40962}
+            (Get-WinEvent -filterHashtable $filter).Count -gt 0 | Should be $true
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Should be,
```Powershell
(Get-WinEvent -filterHashtable $filter).Count | Should BeGreaterThan 0
```

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        }
+        
+        It 'No provider found error' -Skip:(-not $IsWindows) {
+            { New-WinEvent -ProviderName NonExistingProvider -Id 0 } | ShouldBeErrorID 'System.ArgumentException,Microsoft.PowerShell.Commands.NewWinEventCommand'
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Typo. Please change "ShouldBeErrorID" -> "ShouldBeErrorId".

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            { New-WinEvent Microsoft-Windows-PowerShell -Id 999999 } | ShouldBeErrorID 'Microsoft.PowerShell.Commands.EventWriteException,Microsoft.PowerShell.Commands.NewWinEventCommand'
+        }
+
+        It 'IncorrectEventVersion error' -Skip:(-not $IsWindows) {
+            { New-WinEvent -ProviderName Microsoft-Windows-PowerShell -Id 40962 -Version 99 } | ShouldBeErrorID 'Microsoft.PowerShell.Commands.EventWriteException,Microsoft.PowerShell.Commands.NewWinEventCommand'
+        }
+        
+        It 'PayloadMismatch error' -Skip:(-not $IsWindows) {
+            $logPath = join-path $TestDrive 'testlog1.txt'
+            New-WinEvent -ProviderName Microsoft-Windows-PowerShell -Id 32868 *> $logPath # this will print the warning with expected event template to the file
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Please move the comment on separate line.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 (hunk 1 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Yup, I saw that approach; but don't really like it; I think having skip conditions explicitly on each test makes it easier to investigate test failures.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 (hunk 2 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Updated.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 (hunk 3 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Updated.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 (hunk 4 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Moved.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 (hunk 1 above)
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: The recommendation according to the testing guidelines is to use PSDefaultParameterValues. Please see: https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md

Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Updated.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                $PSDefaultParameterValues["it:skip"] = $true
+            }
+        }
+        
+        AfterAll {
+            $global:PSDefaultParameterValues = $originalDefaultParameterValues
+        }
+
+        It 'Simple New-WinEvent' {
+            New-WinEvent -ProviderName Microsoft-Windows-PowerShell -Id 40962 -Version 1 # simple event without any payload
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Move the comment to a separate line.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 (hunk 5 above)
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: use something like

```powershell
$PSProviderId = 40962
```

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
Which issue does this address if any?

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 (hunk 5 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Updated.

**Comment by anmenaga** [comment]
This goes toward #4156

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 (hunk 5 above)
Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: updated.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
Based on the code coverage, you're missing some tests:
https://codecov.io/gh/PowerShell/PowerShell/src/9e41c647a9407e05d86f4971a6bd25719a7130d0/src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventDescriptor.cs
- [ ] EventDescriptor constructor validation like: [System.Diagnostics.Eventing.EventDescriptor]::new(-1, 0, 0, 0, 0, -1, 0)
- [ ] Reading back members of EventDescriptor
https://codecov.io/gh/PowerShell/PowerShell/src/9e41c647a9407e05d86f4971a6bd25719a7130d0/src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProvider.cs
- [ ] Dispose(), Close(), destructor
- [ ] IsEnabled()
- [ ] different data types
- [ ] WriteMessagEvent()
- [ ] WriteEvent
- [ ] WriteTransferEvent
https://codecov.io/gh/PowerShell/PowerShell/src/9e41c647a9407e05d86f4971a6bd25719a7130d0/src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProviderTraceListener.cs
- [ ] EventProviderTraceListener

The Reader code should probably be a different PR

**Comment by anmenaga** [comment]
@SteveL-MSFT All of the things that you've mentioned is ETW code and will go into separate PR.
This PR is specific to New-WinEvent cmdlet code.

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@iSazonov  Can you have a look again?

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            $SimpleEventId = 40962
+            $ComplexEventId = 32868
+        }
+        
+        AfterAll {
+            $global:PSDefaultParameterValues = $originalDefaultParameterValues
+        }
+
+        It 'Simple New-WinEvent' {
+            # simple event without any payload
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: I believe we can remove the comment. And maybe move to the test name.

**Comment by iSazonov** [comment]
With one minor comment LGTM.

File: test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1 (hunk 6 above)
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: @anmenaga can you address this?

Comment (anmenaga) [inline test/powershell/Modules/Microsoft.PowerShell.Diagnostics/New-WinEvent.Tests.ps1]: Updated.

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**